### ISSUE TITLE
compile error: "broken LLVM module found: Operand is null"

### ISSUE THREAD

**Issue by Arwalk** [ziglang/zig#8893]
I've been stuck on a compile error while trying to mix (maybe a bit too much) comptime and runtime variables.

Here is the code that reproduce the error on `0.8.0-dev.2642+75cd37b8f`

```zig
const std = @import("std");

pub const FieldDescriptor = struct {
    tag: u32,
    name: comptime []const u8,
    ftype: u32,
};

pub fn fd(tag: u32, name: []const u8, ftype: u32) FieldDescriptor {
    return FieldDescriptor{
        .tag = tag,
        .name = name,
        .ftype = ftype
    };
}

const FullFieldDescriptor = struct {
    field_name: []const u8,
    tag: u32,
    pb_type: u32,
    real_type: type
};

fn get_struct_field(comptime T: type, comptime field_name: []const u8 ) std.builtin.TypeInfo.StructField {
    return inline for (@typeInfo(T).Struct.fields) |field| {
        if(std.mem.eql(u8, field.name, field_name)) {
            break field;
        }
    } else @compileLog("Could not find field ", field_name, " in struct", T);
}

fn decode(comptime T: type, input : []const u8, allocator: *std.mem.Allocator) !T {
    var result : T = undefined;

    comptime const field_lists = blk: {
        var fields : [T._desc_table.len]FullFieldDescriptor = [_]FullFieldDescriptor{
            .{
                .field_name = "",
                .tag = 0,
                .pb_type = 0,
                .real_type = bool,
            }
        } ** T._desc_table.len;

        inline for (T._desc_table) |field, index| {
            const real_field = get_struct_field(T, field.name);
            fields[index].field_name = field.name;
            fields[index].tag = field.tag;
            fields[index].pb_type = field.ftype;
            fields[index].real_type = real_field.field_type;
        }

        break :blk fields;
    };

    const data_found = blk: {
        var found = std.ArrayList([]const u8).init(allocator);
        var index : usize = 0;
        while (index < input.len) : (index += 2) {
            try found.append(input[index..index+2]);
        }
        break :blk found.toOwnedSlice();
    };
    defer allocator.free(data_found);

    inline for(field_lists) |field| {
        const data = for(data_found) |item| {
            if(field.tag == item[0]) break item;
        } else null;

        if(data) |item| {
            @field(result, field.field_name) = item[1];
        }
    }

    return result;
}

const TestStruct = struct {
    a_field: u32,
    another_field: i32,

    pub const _desc_table = [_]FieldDescriptor{
        fd(1, "a_field", 5),
        fd(2, "another_field", 6),
    };
};

pub fn main() anyerror!void {
    var expected = TestStruct{
        .a_field = 0,
        .another_field = 1
    };

    const encoded = [_]u8{
        1, 0,
        2, 1
    };

    const decoded = try decode(TestStruct, &encoded, std.heap.page_allocator);
    std.log.debug("{any}", .{decoded});
    if(std.meta.eql(expected, decoded)){
        std.log.debug("Zig", .{});
    }
    else
    {
        std.log.debug("Zag", .{});
    }
}

```

This code in general just tries to mix comptime-known values to build some kind of decoder that will then parse input data.
The block starting line 68 (with `inline for(field_lists)...`) Seems to be where the problem lies, as this bug reappeared with this block when i tried to strip down my case to something "simple".

Note that the compiler manages to compile everything if you remove a field in `TestStruct`, and its reference in `_desc_table`. If you comment line 79, 83 and 90, the compiler is fine and the resulting program behave as expected.

Full error:

```
broken LLVM module found: Operand is null
  br i1 %106, label %Then, <null operand!>, !dbg !14684

This is a bug in the Zig compiler.thread 33860 panic:
Unable to dump stack trace: debug info stripped
compilerr...The following command exited with error code 3:
```

I think the numerous inderections i'm doing might cause a problem somewhere.

Edit: Error was reproduced on win_x86_64 and linux_x86_64 (WSL2 environment).

**Comment by ehaas** [comment]
I believe this a bug with comptime type inference; try `const data: ?[]const u8 = for(data_found) |item| {` instead of `const data = for(data_found) |item| {`

**Comment by Arwalk** [comment]
> I believe this a bug with comptime type inference; try `const data: ?[]const u8 = for(data_found) |item| {` instead of `const data = for(data_found) |item| {`

That did the trick, it compiles now. Thanks!

The bug is still present technically though, should i still close this issue?

**Comment by andrewrk** [comment]
Nah keep it open, it will be a nice test case to make sure passes eventually.

**Comment by shmolyneaux** [comment]
I believe that I'm hitting this same bug on `0.9.0-dev.1148+3b09262c1`.

Running `zip run minimal_repro.zig` gives the following output:
```
```
const std = @import("std");
const Allocator = std.mem.Allocator;

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    _ = try get_fn()(&gpa.allocator);
}

fn maybe_value() ?u8 {
    return 3;
}

fn get_fn() fn (*Allocator) anyerror!?void {
    return struct {
        fn inner_fn(allocator: *Allocator) anyerror!?void {
            // NOTE: This needs to be at least 2 elements for the bug to appear
            inline for (.{ 1, 2 }) |_| {
                if (maybe_value()) |_| {
                    _ = try allocator.create(void);
                } else {
                    return null;
                }
            }

            return null;
        }
    }.inner_fn;
}
```

**Comment by shmolyneaux** [comment]
I believe that I'm hitting this same bug with version `0.9.0-dev.1148+3b09262c1`.

Running `zig run minimal_repro.zig` gives the following output:
```
... broken LLVM module found: Operand is null
  br i1 %27, label %ErrRetReturn2, <null operand!>, !dbg !19361

This is a bug in the Zig compiler.thread 12542 panic:
Unable to dump stack trace: debug info stripped
Aborted (core dumped)
```

Here is `minimal_repro.zig`:
```
const std = @import("std");
const Allocator = std.mem.Allocator;

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    _ = try get_fn()(&gpa.allocator);
}

fn maybe_value() ?u8 {
    return 3;
}

fn get_fn() fn (*Allocator) anyerror!?void {
    return struct {
        fn inner_fn(allocator: *Allocator) anyerror!?void {
            // NOTE: This needs to be at least 2 elements for the bug to appear
            inline for (.{ 1, 2 }) |_| {
                if (maybe_value()) |_| {
                    _ = try allocator.create(void);
                } else {
                    return null;
                }
            }

            return null;
        }
    }.inner_fn;
}
```

I'm honestly not sure how I'm supposed to generate functions at compile time, but this was my attempt at it (stripped down from a much larger attempt).

**Comment by brandondyck** [comment]
I think I hit the same bug running this through `zig test` on 0.9.1 for x86_64-windows-gnu:

```zig
const std = @import("std");
const TypeInfo = std.builtin.TypeInfo;

const GenError = error { GenFailed };

pub fn Struct(comptime ValStruct: type) type {
    const struct_info = @typeInfo(ValStruct);
    
    var gen_fields: [struct_info.Struct.fields.len]TypeInfo.StructField = undefined;
    inline for (struct_info.Struct.fields) |field, i| {
        const GenFn = fn () GenError!field.field_type;
        gen_fields[i] = TypeInfo.StructField{
            .name = field.name,
            .field_type = GenFn,
            .default_value = null,
            .is_comptime = false,
            .alignment = @alignOf(GenFn),
        };
    }
    const GenStruct = @Type(TypeInfo{ .Struct = TypeInfo.Struct{
        .layout = .Auto,
        .fields = &gen_fields,
        .decls = &[_]TypeInfo.Declaration{},
        .is_tuple = false,
    } });

    return struct {
        const Self = @This();
        gens: GenStruct,

        pub fn init(gens: GenStruct) Self {
            return .{
                .gens = gens,
            };
        }

        pub fn sample(self: *Self) !ValStruct {
            var value: ValStruct = undefined;
            var last_sampled: u8 = 0;
            const field_infos = @typeInfo(GenStruct).Struct.fields;
            const gens = self.gens;

            errdefer {
                inline for (field_infos) |_, i| {
                    // Adding this var and if breaks the compiler if ValStruct has at least 2 fields:
                    var idx: u8 = i;
                    if (idx > last_sampled) {
                        break;
                    }
                }
            }

            inline for (field_infos) |gen_field, i| {
                const name = gen_field.name;
                @field(value, name) = try @field(gens, name)();
                last_sampled = i;
            }
            return value;
        }
    };
}

fn zero() GenError!u8 {
    return 0;
}

test "Struct.sample" {
    const Ints = struct {a: u8, b: u8};
    var ints = Struct(Ints).init(.{.a = zero, .b= zero});
    _ = try ints.sample();
}
```

Output:
```
broken LLVM module found: Operand is null
  br i1 %35, label %Then, <null operand!>, !dbg !5256

This is a bug in the Zig compiler.thread 4708 panic: 
Unable to dump stack trace: debug info stripped
```

**Comment by hazeycode** [comment]
Another case: https://github.com/hazeycode/clulz/issues/2

**Comment by Vexu** [comment]
Fixed by 4618c41fa6ca70f06c7e65762d2f38d57b00818c
